### PR TITLE
cnf_view with the option to modify clauses

### PR DESCRIPTION
**Pull request by lee30sonia**

### CODE REVIEW

File: include/mockturtle/views/cnf_view.hpp [hunk 1]
```diff
@@ -53,6 +53,94 @@ struct cnf_view_params
 {
   /*! \brief Write DIMACS file, whenever solve is called. */
   std::optional<std::string> write_dimacs{};
+  bool auto_update{true}; /* only used when AllowModify = true */
```
Comment (msoeken) [inline include/mockturtle/views/cnf_view.hpp]: Is this a parameter that should be set by the user? If so, it should be documented (similar to `write_dimacs`)

File: include/mockturtle/views/cnf_view.hpp [hunk 2]
```diff
@@ -53,6 +53,94 @@ struct cnf_view_params
...
+  inline bool is_activated( node const& n ) const
+  {
+    return switch_lit( n ).is_complemented();
+    /* clauses are activated if switch literal is complemented */
+  }
+
+  /*! \brief Deactivates the clauses for a node. */
+  void deactivate( node const& n )
+  {
+    if ( is_activated( n ) )
```
Comment (msoeken) [inline include/mockturtle/views/cnf_view.hpp]: Please use `{ ... }` also around single statement blocks.

File: include/mockturtle/views/cnf_view.hpp [hunk 3]
```diff
@@ -53,6 +53,94 @@ struct cnf_view_params
...
+      switches_[Ntk::node_to_index( n )].complement();
+  }
+
+  /*! \brief (Re-)activates the clauses for a node. */
+  void activate( node const& n )
+  {
+    if ( !is_activated( n ) )
```
Comment (msoeken) [inline include/mockturtle/views/cnf_view.hpp]: Please use `{ ... }` also around single statement blocks.

File: include/mockturtle/views/cnf_view.hpp [hunk 4]
```diff
@@ -53,6 +53,94 @@ struct cnf_view_params
...
   std::optional<std::string> write_dimacs{};
+  bool auto_update{true}; /* only used when AllowModify = true */
+};
+
+/* forward declaration */
+template<typename Ntk, bool AllowModify = false, bill::solvers Solver = bill::solvers::glucose_41>
+class cnf_view;
+
+template<typename CnfView, typename Ntk, bool AllowModify = false, bill::solvers Solver = bill::solvers::glucose_41>
+class cnf_view_impl : public Ntk
```
Comment (msoeken) [inline include/mockturtle/views/cnf_view.hpp]: This class should be put into a `detail` namespace.

File: include/mockturtle/views/cnf_view.hpp [hunk 5]
```diff
@@ -115,6 +222,8 @@ class cnf_view : public Ntk
   /* \brief Returns the variable associated to a node. */
   inline bill::var_type var( node const& n ) const
   {
+    if constexpr ( AllowModify )
+      return cnf_view_impl_t::literals_[n].variable();;
```
Comment (msoeken) [inline include/mockturtle/views/cnf_view.hpp]: There is an extra `;`

File: include/mockturtle/views/cnf_view.hpp [hunk 6]
```diff
@@ -115,6 +222,8 @@ class cnf_view : public Ntk
   /* \brief Returns the variable associated to a node. */
   inline bill::var_type var( node const& n ) const
   {
+    if constexpr ( AllowModify )
```
Comment (msoeken) [inline include/mockturtle/views/cnf_view.hpp]: Please use `{ ... }` also around single statement blocks.

File: include/mockturtle/views/cnf_view.hpp [hunk 7]
```diff
@@ -275,30 +424,77 @@ class cnf_view : public Ntk
...
+
+    if constexpr ( AllowModify )
+    {
+      if ( add_var )
+      {
+        node_lit = bill::lit_type( solver_.add_variable(), bill::lit_type::polarities::positive );
+        cnf_view_impl_t::literals_.resize( bill::lit_type(0, bill::lit_type::polarities::positive) );
+        cnf_view_impl_t::literals_[n] = node_lit;
+      }
+      else
```
Comment (msoeken) [inline include/mockturtle/views/cnf_view.hpp]: Please use `{ ... }` also around single statement blocks.

File: lib/bill/bill/sat/interface/types.hpp [hunk 8]
```diff
@@ -23,7 +23,7 @@ class var_type {
 	constexpr static uint32_t max_value = (std::numeric_limits<uint32_t>::max() >> 1);
 
 public:
-	constexpr var_type(uint32_t var)
+	constexpr var_type(uint32_t var = 0)
```
Comment (msoeken) [inline lib/bill/bill/sat/interface/types.hpp]: We need to make sure that this is back-ported to bill.

File: include/mockturtle/views/cnf_view.hpp [hunk 9]
```diff
@@ -71,18 +159,22 @@ struct cnf_view_params
  * added.  Therefore, a network cannot modify or delete nodes when wrapped in a
  * `cnf_view`.
  */
-template<typename Ntk, bill::solvers Solver = bill::solvers::glucose_41>
-class cnf_view : public Ntk
+template<typename Ntk, bool AllowModify, bill::solvers Solver>
```
Comment (msoeken) [inline include/mockturtle/views/cnf_view.hpp]: We need some tests in `cnf_view.cpp` for the case that `AllowModify = true` starting from an existing network, modified network, deleted node, ...

File: lib/bill/bill/sat/interface/types.hpp (hunk 8 above)
Comment (lee30sonia) [reply lib/bill/bill/sat/interface/types.hpp]: Yeah the changes here are exactly the same as my PR to bill.

**Pull request merged**